pkg/server/controller.go

```diff
@@ -145,6 +282,264 @@ type indexPage struct {
+		requestState := r.FormValue("state")
+
+		if requestState != cookieState {
+			ctrl.httpServer.ErrorLog.Printf("invalid oauth state, expected %v got %v", cookieState, requestState)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		code := r.FormValue("code")
+		if code == "" {
+			ctrl.httpServer.ErrorLog.Printf("Code not found")
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		token, err := oauthConf.Exchange(oauth2.NoContext, code)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Exchanging auth code for token failed with %v ", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		client := oauthConf.Client(oauth2.NoContext, token)
+		resp, err := client.Get(getAccountInfoURL)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Failed to get oauth user info: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+		defer resp.Body.Close()
+
+		name, err := decodeResponse(resp)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Decoding response body failed: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		claims := jwt.MapClaims{
+			"iat":  time.Now().Unix(),
+			"name": name,
+		}
+
+		if ctrl.config.LoginMaximumLifetimeDays > 0 {
+			claims["exp"] = time.Now().Add(time.Hour * 24 * time.Duration(ctrl.config.LoginMaximumLifetimeDays)).Unix()
+		}
+
+		jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+		tk, err := jwtToken.SignedString([]byte(ctrl.config.JWTSecret))
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Signing jwt failed: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		// delete state cookie and add jwt cookie
+		stateCookie := &http.Cookie{
+			Name:     stateCookieName,
+			Path:     "/",
+			Value:    "",
+			HttpOnly: true,
+			// MaxAge -1 request cookie be deleted immediately
+			MaxAge:   -1,
+			SameSite: http.SameSiteStrictMode,
+		}
+
+		http.SetCookie(w, stateCookie)
+
+		jwtCookie := &http.Cookie{
+			Name:     jwtCookieName,
+			Path:     "/",
+			Value:    tk,
+			HttpOnly: true,
+			MaxAge:   0,
+			SameSite: http.SameSiteStrictMode,
+		}
+
+		http.SetCookie(w, jwtCookie)
+		tmplt := template.New("welcome.html")
+		tmplt, _ = tmplt.ParseFiles("./webapp/templates/welcome.html")
+		params := map[string]string{"Name": name}
+
+		tmplt.Execute(w, params)
+		return
+	}
+}
+
+// can be replaced with a faster solution if cryptographic randomness isn't a priority
+func generateStateToken(length int) (string, error) {
+	b := make([]byte, length)
+	if _, err := rand.Read(b); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(b), nil
+}
+
+func (ctrl *Controller) oauthLoginHandler(oauthConf *oauth2.Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		URL, err := url.Parse(oauthConf.Endpoint.AuthURL)
```

Comment:
We should perform all the validation steps before creating `oauth2.Config` instance. Consider use of shallow copy (https://github.com/golang/go/issues/41733).

 `URL` would be `authUrl` (`url` collides to the package name).

Reply:
I'm not sure I understand what should I use shallow copy for? Also if we were to call url.Parse before building oauth2.Config we would have to either pass some sort of identifier and the create it in a switch or remove url parsing to above function which would triple the amount of code. URL parsing should rarely fail so it won't influence performance that much that we created instance before if it fails.

Or did I misunderstand something?

Reply:
I agree. My point is that when we create `*oauth2.Config` we should validate user input (fail fast).

BTW, am I right that scopes are passed as a space delimited string? I'm wondering if it works - CLI may not interpret that in the right way.

Next are just my subjective thought :) Given that we already have parsed URL at validation, should we do that again for every request? For `oauthLoginHandler` we use `*oauth2.Config` as a data container - instead we could introduce our own type to bear prepared data. But it's up to you ;)

Reply:
Removed scope as config flag so that won't be a problem :) 

I really like the idea. Will include this.